### PR TITLE
Fix the elusive nginx 502 error

### LINKED ISSUE #3546: SetCookie headers jetpackState results in 500 errors with Nginx in request path (including CloudFlare)
#### What I expected

JetPack installation works. 
#### What happened instead

On final authorization, an error is delivered from the webserver (not WordPress).

Nginx and Varnish users have both reported issues that appear to be occurring due to the number of `SetCookie: jetpackState[some_state]` headers. These headers duplicate `SetCookie` directives for a number of `some_state` values, meaning they will not be reliably interpreted by browsers, and are likely not necessary.

For example, this is the root cause of this failure under roots.io's nginx setup: h

https://github.com/roots/trellis/issues/323

It's also easily replicable on Pantheon (where I have an interest in things working smoothly.) It occurs when installing JetPack fresh, and can also be replicated by bulk-enabling a large number of features. 
#### Steps to reproduce the issue

Install JetPack on a fresh WP site on Pantheon (it's free!). You'll get a 503 error on activation at the end. Captured headers look like:

https://gist.github.com/joshkoenig/7c436c4f158b106eb8e2

Disabling and re-enabling will replicate. You can also replicate by bulk-enabling features. 

However, JetPack installation is ultimately successful (e.g. if you releoad it works).  We're looking into why this loop is happening, but generally speaking on the JetPack side the setting of these state cookies could probably be smarter.

### PR DESCRIPTION
**Pull request by withinboredom**
Fixes #3546

#### Changes proposed in this Pull Request:

* After talking with @roccotripaldi and @kraftbj (100% @kraftbj's idea btw, I just executed) ... we don't need to set state on xmlrpc requests. So, let's not do that. Now we're not trying to send a bajillion length headers back :D

#### Testing instructions:

* nginx default config will no longer respond with 502 during authorization

#### Proposed changelog entry for your changes:

- Fix elusive nginx 502 error...

### CODE REVIEW

File: class.jetpack.php [hunk 1]
```diff
@@ -5228,6 +5228,9 @@ public static function clean_nonces( $all = false ) {
 	 * @param bool $restate private
 	 */
 	public static function state( $key = null, $value = null, $restate = false ) {
+		if ( XMLRPC_REQUEST ) {
```
Comment (kraftbj) [inline class.jetpack.php]: Should be either `defined( 'XMLRPC_REQUEST' )` or `defined('XMLRPC_REQUEST') && XMLRPC_REQUEST )`

**Comment by zinigor** [comment]
Sorry gents, but disabling state in this case seems a bit over the top. I have pushed a commit that silences `activate_default_modules` when calling the authorize method, that fixes the problem for me.

**Comment by zinigor** [comment]
For 5.1 I'd suggest returning to this problem and leave module activation until the first time the user visits wp-admin.

**Comment by roccotripaldi** [comment]
> Sorry gents, but disabling state in this case seems a bit over the top.

Can you provide some examples of when we'd need state during XMLRPC?

> For 5.1 I'd suggest returning to this problem and leave module activation until the first time the user visits wp-admin.

I think we are activating default modules during authorization because we needed access immediately to jetpack manage ( no longer an issue ), and json api

**Comment by zinigor** [comment]
> Can you provide some examples of when we'd need state during XMLRPC?

We most certainly don't need to set state when issuing XMLRPC requests between WordPress.com and Jetpack sites, but in this situation that is not the case. When a user click the "Approve" button in Calypso, it's the user's browser that's issuing the request, not WordPress.com. And though we don't use any state messages received this way, we may need to add messages to state anyway, at least to preserve them for next wp-admin screen loads.

> I think we are activating default modules during authorization because we needed access immediately to jetpack manage ( no longer an issue ), and json api

Yes, good point, I forgot about that.

**Comment by kraftbj** [comment]
> When a user click the "Approve" button in Calypso, it's the user's browser that's issuing the request, not WordPress.com.

Ah, I didn't realize that was client->JP-site-server and not just WP.com<->JP-site-server. Makes sense to keep the ability to set/read state in that case.

The solution may be some type of check to ensure that the state doesn't get out of hand with excessive cookie setting? I think disabling the state calls in this particular instance (default modules) makes sense, but there could be other situations where we get out of hand (the OP's reference of the bulk-activation of many modules via the old modules screen, even though that particular instance is outdated).

**Comment by zinigor** [comment]
Yes, I was looking at refactoring state storage because right now it just calls `setcookie` on a first come first serve basis. So if you have several messages that overwrite each other, they will all get output, not just the latest one.

Fixing that will definitely improve the situation for all cases where this bug shows itself, not only for authorize calls. But I thought this was going to be a 5.0.1 sure-fire change, while for 5.1 we can look at something more solid. As to if there will be a 5.0.1 - I'm not sure :)

**Comment by kraftbj** [comment]
👍 Makes sense. I think if we can get this in for at least the next version, that is the minimal need. The whole refactor isn't a must-land for 5.1, but obviously would be good to have sooner rather than later.

**Comment by kraftbj** [comment]
@withinboredom @roccotripaldi Are y'all okay with @zinigor's change of direction on this? We can still audit the state in the future, but not get held up on that to ensure this makes it into the July release.

**Pull request merged**